pkg/virtualization/core/virtualsystem/virtualmachine.go

```diff
@@ -328,6 +346,129 @@ func (vm *VirtualMachine) GetVirtualSystemSettingData() (*VirtualSystemSettingDa
 	return NewVirtualSystemSettingData(inst)
 }
 
+func (vm *VirtualMachine) GetVirtualGuestNetworkAdapterConfiguration(inputMacAddress string) (guestNetworkAdapterConfiguration *na.GuestNetworkAdapterConfiguration, err error) {
+	allSettings, err := vm.GetRelatedEx("Msvm_SettingsDefineState", "Msvm_VirtualSystemSettingData", "", "")
+	if err != nil {
+		return nil, err
+	}
+
+	/* The input MAC address would be always in standard format (i.e aa:bb:cc:dd:ee:ff)
+	   But the address read from system would be in HyperV format (i.e AABBCCDDEEFF) */
+	inputMacAddressHyperV := strings.ReplaceAll(inputMacAddress, ":", "")
+	inputMacAddressHyperV = strings.ReplaceAll(inputMacAddressHyperV, "-", "")
+	inputMacAddressHyperV = strings.ToUpper(inputMacAddressHyperV)
+
+	for _, settings := range allSettings {
+		wmiSyntheticNetworkAdapters, err := settings.GetAllRelated("Msvm_SyntheticEthernetPortSettingData")
+		if err != nil {
+			continue
+		}
+
+		for _, wmiSyntheticNetworkAdapter := range wmiSyntheticNetworkAdapters {
+
+			syntheticNetworkAdapter, err := na.NewSyntheticNetworkAdapter(wmiSyntheticNetworkAdapter)
+			if err != nil {
+				continue
+			}
+
+			networkAdapterMacAddress, err := syntheticNetworkAdapter.GetPropertyAddress()
+			if err != nil {
+				continue
+			}
+
+			if strings.EqualFold(inputMacAddressHyperV, networkAdapterMacAddress) {
+				wmiGuestConfig, err := syntheticNetworkAdapter.GetRelated("Msvm_GuestNetworkAdapterConfiguration")
+				if err != nil {
+					continue
+				}
+				guestNetworkAdapterConfiguration, _ = na.NewGuestNetworkAdapterConfiguration(wmiGuestConfig)
+				return guestNetworkAdapterConfiguration, nil
+			}
+		}
+	}
+
+	return nil, errors.Wrapf(err, "Unable to get the network adapter configuration for the vnic with mac address [%s] on vm [%s]", inputMacAddress, vm.ElementName)
+}
+
+func (vm *VirtualMachine) GetSecuritySettingData() (value *MsvmSecuritySettingData, err error) {
+	inst, err := vm.GetRelated("Msvm_Tpm")
+
+	// If the TPM is not found, then it is not configured or enabled
+	if inst == nil {
+		return nil, nil
+	}
+
+	if err != nil {
```

Comment:
We should probably move this condition above `if inst == nil` on line 397 since I'd assume `inst` will also be nil in the error case. Otherwise I think we'll always return on line 398 if this fails and would lose the error information.